Fixed terminator deserialisation, and fixed encoding type.

CitadelCore opened the pull request:
One problem I found is that the binary deserialiser does not read the terminators at the end of the document in a stream. This is fine when you have one file with plain binary keyvalues1, but when you have a file with multiple sections with custom headers (like `appinfo.vdf`), you run into problems when reading the next section's header because there are still terminator bytes that haven't been read.

Additionally, keys are saved as ASCII and values as UTF-8. If the file contains UTF-8 data (like the multilingual content in `appinfo.vdf`) there is the potential for data loss.

I've implemented a fix for both of these problems - the former, I've modified `KV1BinarySerializer.cs` to read all terminator bytes at the end, and the latter, I've modified both `KV1BinarySerializer.cs` and `KV1BinaryReader.cs` to read UTF-8 rather than ASCII in object string values.

All unit tests pass. Hope these improvements will prove useful.

xPaw commented on ValveKeyValue/ValveKeyValue/Deserialization/KV1BinaryReader.cs:
```diff
@@ -17,19 +17,19 @@ public KV1BinaryReader(Stream stream, IVisitationListener listener)
-        readonly BinaryReader reader;
-        readonly IVisitationListener listener;
-        bool disposed;
+        private readonly Stream _stream;
```
Can you revert these changes please?

It makes for an unnecessarily hard diff to read.

CitadelCore replied:
Done

yaakov-h commented on ValveKeyValue/ValveKeyValue/Deserialization/KV1BinaryReader.cs:
```diff
@@ -61,82 +61,81 @@ public void Dispose()
         }
 
-        string ReadNullTerminatedString()
+        string ReadNullTerminatedString(Encoding encoding = null)
```
This seems misleading to me because it only works with encoding that are terminated by a single NUL byte. I'm not 100% sure this actually covers UTF-8 correctly, but it definitely doesn't work with UTF-16 or UTF-32. Perhaps this function should just read the null terminated byte stream, and the caller should decode.

CitadelCore replied:
Moved to a function that just reads the null-terminated byte array, and moved the actual decoding to the callers.

yaakov-h commented on ValveKeyValue/ValveKeyValue/Deserialization/KV1BinaryReader.cs:
```diff
@@ -61,82 +61,81 @@ public void Dispose()
+                b = reader.ReadByte();
+
+            reader.BaseStream.Position -= len + 1;
+            var str = reader.ReadBytes(len);
```
We don't need to read twice. Consider keeping a buffer, and resizing it if we overrun.

CitadelCore replied:
Now using MemoryStream.

yaakov-h commented on ValveKeyValue/ValveKeyValue/Serialization/KV1BinarySerializer.cs:
```diff
@@ -78,13 +78,12 @@ void Write(KV1BinaryNodeType nodeType)
-            {
-                writer.Write(@char);
-            }
+            if (encoding == null) encoding = Encoding.ASCII;
```
As above regarding NUL termination, but we don't really have a better option here.

CitadelCore replied:
Modified to write a terminated byte array instead of a string and moved encoding to the caller.

yaakov-h commented on ValveKeyValue/ValveKeyValue/Deserialization/KV1BinaryReader.cs:
```diff
@@ -61,82 +61,81 @@ public void Dispose()
-            KVValue value;
-
-            switch (type)
+            while (true)
```
I need to check the larger structure, but looping here seems wrong. We're trying to read one object, not many. If we're just not reading `End` bytes then we should add that, as a peek if necessary.

CitadelCore replied:
I've removed the loop and moved logic back to ReadObjectCore and ReadValue, which works fine.

CitadelCore commented:
As well as the changes in the comments, I've added UTF-8 tests with Chinese characters to `BinaryObjectSerializationTestCase` and `BinaryObjectDeserializationTestCase.` I've also added two new unit tests, `BinaryObjectConsecutiveSerializationTestCase` and `BinaryObjectConsecutiveDeserializationTestCase`, for consecutive serialisation and deserialisation.

I've added the class `NonClosingStream` as a test helper, since the deserialiser will automatically close the stream when it finishes, which is not what we want when consecutively reading objects. Perhaps there should be an option to leave the stream open like the one in `BinaryWriter`, which would avoid this issue entirely.

CitadelCore commented on ValveKeyValue/ValveKeyValue.Test/Helpers/NonClosingStream.cs:
```diff
@@ -0,0 +1,311 @@
+﻿using System;
+using System.IO;
+
+namespace CSERLibrary.Utilities
```
Woops - I copied it from my own project, which in itself obtained it from [www.yoda.archsys.com/csharp/miscutil/](url). I've fixed the namespace. The original code is under the Apache license.

xPaw replied:
`NonClosingStream` will not be needed if you add `leaveOpen: true` where it is missing as @yaakov-h suggested instead.

CitadelCore replied:
Fixed. KV1BinaryReader both didn't have `leaveOpen: true` in the BinaryReader constructor and additionally disposed the stream manually in its Dispose method.

xPaw commented on ValveKeyValue/ValveKeyValue/Deserialization/KV1BinaryReader.cs:
```diff
@@ -54,79 +54,79 @@ public void Dispose()
+                listener.OnObjectStart(name);
+                ReadObjectCore();
+                listener.OnObjectEnd();
+            }
```
If you `return` in the if, then you won't need to nest everything in the `else`.

CitadelCore replied:
Don't see why this can't be done? You've struck it through.

CitadelCore replied:
Implemented @yaakov-h's suggestion.

xPaw commented on ValveKeyValue/ValveKeyValue/Deserialization/KV1BinaryReader.cs:
```diff
@@ -54,79 +54,79 @@ public void Dispose()
+
+            // Keep reading values, until we reach the terminator
+            while ((type = ReadNextNodeType()) != KV1BinaryNodeType.End)
+                ReadValue(type);
```
Add braces please (and in other places too)

CitadelCore replied:
Changed.

CitadelCore commented on ValveKeyValue/ValveKeyValue/Deserialization/KV1BinaryReader.cs:
```diff
@@ -54,79 +54,79 @@ public void Dispose()
-            KVValue value;
-
-            switch (type)
+            // ASCII encoding is used for keys
```
I just did some testing and actually, there seems to be no problems using UTF8 for the keys,

xPaw replied:
@yaakov-h what do you think about this?

CitadelCore replied:
Changed to UTF8.

yaakov-h replied:
@xPaw if only there was an actual spec...

xPaw commented on ValveKeyValue/ValveKeyValue.Test/Binary/BinaryObjectConsecutiveDeserializationTestCase.cs:
```diff
@@ -0,0 +1,68 @@
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using Microsoft.VisualStudio.TestPlatform.ObjectModel;
```
What is this used by? From a quick glance I can't see it, but I haven't fired up VS to check.

CitadelCore replied:
A garbage using that was automatically inserted by resharper when I copied over some of my own code. Removed it.

xPaw replied:
Ah.

yaakov-h replied:
It’s still there, though?
